src/datasets/fingerprint.py

```diff
@@ -381,7 +381,7 @@ def wrapper(*args, **kwargs):
                     kwargs_for_fingerprint.pop(default_varname)
 
             # compute new_fingerprint and add it to the args of not in-place transforms
-            transform = func.__module__ + "." + func.__qualname__
+            transform = func.__module__ + "." + func.__name__
```

Comment:
This is probably the weirdest thing about this PR. Upon investigating, the fingerprint discrepancy between the mutiprocess version and the sequential one, I noticed that `__qualname__` had different values:
 - multiprocessing: `_map_single`
 - sequential: `Dataset._map_single`
 
 I tried figuring out where this behaviour was defined in `multiprocess` but I couldn't seem to find that. Instead, I'm not sure we actually need `__qualname__` as this seems to apply only to internal HF functions. Instead I suggest we replace it with `__name__` in order to fix this discrepancy, the multiprocessing version remains unchanged.
 
 Better if you know where I can try to fix this issue in `multiprocess`, I'd be more than happy to revert.

Reply:
I'm not able to reproduce this:
```python
from multiprocess import Pool

def func_name():
    import datasets
    print(datasets.Dataset._map_single.__name__)

def func_qualname():
    import datasets
    print(datasets.Dataset._map_single.__qualname__)

func_name()
with Pool(2) as p:
    p.apply_async(func_name).get()
# _map_single
# _map_single

func_qualname()
with Pool(2) as p:
    p.apply_async(func_qualname).get()
# Dataset._map_single
# Dataset._map_single
```
(tested on colab and on my macbook pro)

Therefore I would leave it as before, since `__name__` could create collisions in the names in the future.

Could you provide a reproducible example ?

Reply:
```python
from functools import wraps

def print_func_name(func):
  @wraps(func)
  def wrapper(*args, **kwargs):
    print(func.__qualname__, func.__name__)
  return wrapper

class Foo:
  @print_func_name
  def bar():
    pass

func_qualname()
with Pool(2) as p:
    p.apply_async(Foo.bar).get()
```

Reply:
I was able to find a way to bypass this issue, but it really feels like a hack ...

```python
from functools import wraps
from multiprocess import Pool
def print_func_name(func):
  temp = func.__qualname__
  def wrapper(*args, **kwargs):
    print(func.__qualname__, func.__name__, temp)
  return wrapper

class Foo:
  @print_func_name
  def bar():
    pass

Foo.bar()
with Pool(2) as p:
    p.apply_async(Foo.bar).get()
    ```

Reply:
Nice ! I added a test and a comment to explain this, so I think it's fine :)